### PR TITLE
ddl:set the low priority for the adding the index

### PR DESCRIPTION
**Pull request by ciscoxll**

### CODE REVIEW

File: domain/domain.go [hunk 1]
```diff
@@ -67,7 +67,7 @@ type Domain struct {
 // It returns the latest schema version, the changed table IDs, whether it's a full load and an error.
 func (do *Domain) loadInfoSchema(handle *infoschema.Handle, usedSchemaVersion int64, startTS uint64) (int64, []int64, bool, error) {
 	var fullLoad bool
-	snapshot, err := do.store.GetSnapshot(kv.NewVersion(startTS))
+	snapshot, err := do.store.GetSnapshotWithPriority(kv.NewVersion(startTS), kv.PriorityHigh)
```
Comment (shenli) [inline domain/domain.go]: All the kv operations for meta data will use high-priority. So we do not need to set it here.

**Comment by shenli** [comment]
Please set the priority [here](https://github.com/pingcap/tidb/blob/master/ddl/index.go#L652).

File: domain/domain.go (hunk 1 above)
Comment (ciscoxll) [reply domain/domain.go]: Already remove

**Comment by ciscoxll** [comment]
@zimulala  PTAL

File: kv/txn.go [hunk 2]
```diff
@@ -32,6 +32,7 @@ func RunInNewTxn(store Storage, retryable bool, f func(txn Transaction) error) e
 		txn           Transaction
 	)
 	for i := uint(0); i < maxRetryCnt; i++ {
+		store.SetPriority(PriorityLow)
```
Comment (zimulala) [inline kv/txn.go]: Setting priority here will affect other operations that uses RunInNewTxn.

Comment (ciscoxll) [reply kv/txn.go]: Ok, I changed.

**Comment by ciscoxll** [comment]
@zimulala  @winkyao PTAL

File: ddl/index.go [hunk 3]
```diff
@@ -654,7 +654,9 @@ func (w *worker) doBackfillIndexTask(t table.Table, colMap map[int64]*types.Fiel
 
 	startTime := time.Now()
 	var ret *taskResult
-	err := kv.RunInNewTxn(w.ctx.GetStore(), true, func(txn kv.Transaction) error {
+	store := w.ctx.GetStore()
+	store.SetPriority(kv.PriorityLow)
```
Comment (shenli) [inline ddl/index.go]: Will this affect the global priority of the store instance.

**Comment by zimulala** [comment]
Cloud we add some tests like `executor/executor_test.go` in #4119？

File: ddl/index.go (hunk 3 above)
Comment (zimulala) [reply ddl/index.go]: Like https://github.com/pingcap/tidb/blob/master/meta/meta.go#L89

Comment (winkyao) [reply ddl/index.go]: set txn.GetSnapshot() priority is ok.

File: ddl/column.go [hunk 4]
```diff
@@ -304,7 +304,7 @@ func (d *ddl) addTableColumn(t table.Table, columnInfo *model.ColumnInfo, reorgI
 	for {
 		startTime := time.Now()
 		handles = handles[:0]
-		err = iterateSnapshotRows(d.store, t, version, seekHandle,
+		err = iterateSnapshotRows(d.store, kv.PriorityNormal, t, version, seekHandle,
```
Comment (winkyao) [inline ddl/column.go]: no need to add priority param, set txn priority like zimulala saied.

File: plan/logical_plan_test.go [hunk 5]
```diff
@@ -359,6 +359,10 @@ func (m *mockStore) GetSnapshot(ver kv.Version) (kv.Snapshot, error) {
 	return nil, nil
 }
 
+func (m *mockStore) SetPriority(priority int) {
```
Comment (coocood) [inline plan/logical_plan_test.go]: This can be removed.

File: kv/mock.go [hunk 6]
```diff
@@ -148,6 +148,10 @@ func (s *mockStorage) Close() error {
 	return nil
 }
 
+func (s *mockStorage) SetPriority(priority int) {
```
Comment (coocood) [inline kv/mock.go]: This can be removed.

File: ddl/index.go [hunk 7]
```diff
@@ -715,13 +716,12 @@ func allocateIndexID(tblInfo *model.TableInfo) int64 {
 // recordIterFunc is used for low-level record iteration.
 type recordIterFunc func(h int64, rowKey kv.Key, rawRecord []byte) (more bool, err error)
 
-func iterateSnapshotRows(store kv.Storage, t table.Table, version uint64, seekHandle int64, fn recordIterFunc) error {
+func iterateSnapshotRows(store kv.Storage, priority int, t table.Table, version uint64, seekHandle int64, fn recordIterFunc) error {
```
Comment (coocood) [inline ddl/index.go]: The priority is not used.

File: store/tikv/snapshot.go [hunk 8]
```diff
@@ -38,7 +38,8 @@ const (
 )
 
 // tikvSnapshot implements the kv.Snapshot interface.
-type tikvSnapshot struct {
+type
+tikvSnapshot struct {
```
Comment (winkyao) [inline store/tikv/snapshot.go]: remove `enter`

File: kv/kv.go [hunk 9]
```diff
@@ -220,6 +220,8 @@ type Snapshot interface {
 	Retriever
 	// BatchGet gets a batch of values from snapshot.
 	BatchGet(keys []Key) (map[string][]byte, error)
+	// SetPriority snapshot set the priority
+	SetPriority(priority int)
```
Comment (tiancaiamao) [inline kv/kv.go]: I dislike this change.

File: ddl/index.go [hunk 10]
```diff
@@ -715,13 +716,13 @@ func allocateIndexID(tblInfo *model.TableInfo) int64 {
...
 	ver := kv.Version{Ver: version}
 	snap, err := store.GetSnapshot(ver)
+	snap.SetPriority(priority)
```
Comment (tiancaiamao) [inline ddl/index.go]: That's too complex....
The easiest change:

```
txn := store.BeginWithStartTs(version)
txn.SetOption(kv.Priority, kv.PriorityLow)
snap := txn.GetSnapshot()
```

File: store/tikv/txn.go [hunk 11]
```diff
@@ -56,6 +56,7 @@ func newTiKVTxn(store *tikvStore) (*tikvTxn, error) {
 func newTikvTxnWithStartTS(store *tikvStore, startTS uint64) (*tikvTxn, error) {
 	ver := kv.NewVersion(startTS)
 	snapshot := newTiKVSnapshot(store, ver)
+	snapshot.priority = kvPriorityToCommandPri(store.priority)
```
Comment (winkyao) [inline store/tikv/txn.go]: remove this line.

File: ddl/index.go [hunk 12]
```diff
@@ -715,13 +716,13 @@ func allocateIndexID(tblInfo *model.TableInfo) int64 {
 // recordIterFunc is used for low-level record iteration.
 type recordIterFunc func(h int64, rowKey kv.Key, rawRecord []byte) (more bool, err error)
 
-func iterateSnapshotRows(store kv.Storage, t table.Table, version uint64, seekHandle int64, fn recordIterFunc) error {
+func iterateSnapshotRows(store kv.Storage,t table.Table, version uint64, seekHandle int64, fn recordIterFunc) error {
 	ver := kv.Version{Ver: version}
 	snap, err := store.GetSnapshot(ver)
+	snap.SetPriority(kv.PriorityLow)
```
Comment (tiancaiamao) [inline ddl/index.go]:
```
txn := store.BeginTxnWithStartTS(version)
txn.SetOption(kv.Priority, kv.LowPriority)
snap := txn.GetSnapshot()
```

File: util/arena/arena.go [hunk 13]
```diff
@@ -60,7 +60,7 @@ func NewAllocator(capacity int) *SimpleAllocator {
 // Alloc implements Allocator.AllocBytes interface.
 func (s *SimpleAllocator) Alloc(capacity int) []byte {
 	if s.off+capacity < cap(s.arena) {
-		slice := s.arena[s.off : s.off : s.off+capacity]
+		slice := s.arena[s.off:s.off : s.off+capacity]
```
Comment (tiancaiamao) [inline util/arena/arena.go]: Update your latest Go to the latest version

**Comment by winkyao** [comment]
@coocood PTAL

File: store/tikv/kv.go [hunk 14]
```diff
@@ -181,6 +182,7 @@ func newTikvStore(uuid string, pdClient pd.Client, spkv SafePointKV, client Clie
 		safePoint:   0,
 		spTime:      time.Now(),
 		closed:      make(chan struct{}),
+		priority:    kv.PriorityNormal,
```
Comment (coocood) [inline store/tikv/kv.go]: This can be removed.

File: store/tikv/kv.go [hunk 15]
```diff
@@ -138,6 +138,7 @@ type tikvStore struct {
 	spTime    time.Time
 	spMutex   sync.RWMutex  // this is used to update safePoint and spTime
 	closed    chan struct{} // this is used to nofity when the store is closed
+	priority  int
```
Comment (coocood) [inline store/tikv/kv.go]: This can be removed.

File: executor/executor_test.go [hunk 16]
```diff
@@ -2071,6 +2071,31 @@ func (s *testContextOptionSuite) TearDownSuite(c *C) {
...
+	}
+
+	cli := s.cli
+	cli.mu.Lock()
+	cli.mu.checkFlags = checkRequestPriority
+	cli.mu.Unlock()
+
+	cli.priority = pb.CommandPri_Low
+	tk.MustQuery("alter table t1 add index t1_index (id);")
+	tk.MustQuery("select * from t1 where id = 1")
```
Comment (shenli) [inline executor/executor_test.go]: So the `select * from t1 where id = 1` will use low priority?

Comment (ciscoxll) [reply executor/executor_test.go]: already remove

**Comment by coocood** [comment]
@ciscoxll 
The test can count the number of Low Priority Request in the checkRequestClient, and assert it is greater than a value.

**Comment by coocood** [comment]
LGTM

**Comment by shenli** [comment]
LGTM

**Comment by ciscoxll** [comment]
# TiDB Sysbench
## Testing environment
### Hardware Config
* Intel Xeon E312xx 16 cores 2.4Ghz.
* 16G Memory 1600MHz .
* SSD 2.0GB+/s.
*  1 node.

**Pull request merged**